src/main/auth/o-auth2.ts

```diff
@@ -0,0 +1,149 @@
+import assert from 'assert';
+import { URLSearchParams } from 'url';
+import { Request } from '../request';
+import { RequestAuthorization, AuthRetryConfig } from '../definitions';
+
+export enum OAuth2GrantType {
+    CLIENT_CREDENTIALS = 'client_credentials',
+    REFRESH_TOKEN = 'refresh_token',
+}
+
+export interface OAuth2Params {
+    // required if we want to enforce refreshing..
+    clientId: string;
+    tokenUrl: string;
+
+    // optional
+    clientSecret?: string;
+    refreshToken?: string;
+    accessToken?: string;
+    expiresAt?: number;
+}
+
+export class OAuth2 implements RequestAuthorization, OAuth2Params {
```

Comment:
This q applies to both auth implementations: why do we want to implement `OAuth2Params` interface? Isn't it easier to just store the auth params separately here and just `this.params = { ...params }` in constructor?

Reply:
I wanted to avoid typing extra `this.params.XXX` in other methods 😁

Reply:
Lol I see, but probs still worth isolating to keep each RequestAuth interface leaner (i.e. since we already plan to access some of them directly). Remember: interfaces are harder to change once consumers start using them :)